# Credit Risk Analysis

Projeto desenvolvido por:
* Mariana Ramos - up201806869
* Pedro Ferreira - up201806506
* Pedro Ponte - up201809694

In order to be able to make conclusions and reach our goal, we will need to extract and study the data. To do so, we have to import the panda library and extract the information to be stored in a variable so we can work with it. 

In [78]:
import pandas as pd

credit_data = pd.read_csv('data.csv', na_values=['NA'], low_memory=False)
credit_data.shape

(855969, 73)

As we can see the data file has a total of 855969 different results and a total of 73 evaluation criteria. The second parameter of the read_csv function makes all the empty fields be filled with NA so we can easily analyse missing values in the future.

We can start by reading the data into a pandas DataFrame to see if is everything alright.

In [41]:
credit_data.head(10)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   
5  1075269    1311441       5000         5000           5000.0   36 months   
6  1069639    1304742       7000         7000           7000.0   60 months   
7  1072053    1288686       3000         3000           3000.0   36 months   
8  1071795    1306957       5600         5600           5600.0   60 months   
9  1071570    1306721       5375         5375           5350.0   60 months   

   int_rate  installment grade sub_grade  ... il_util open_rv_12m open_rv_24m  \
0     10.65       162.87     B        B2  ...     NaN         NaN         NaN   
1     15.27        59.83     C        C4  ...     NaN         NaN         NaN   
2     15.96        84.33     C        C5  ...     NaN         NaN         NaN   
3     13.49       339.31     C        C1  ...     NaN         NaN         NaN   
4     12.69        67.79     B        B5  ...     NaN         NaN         NaN   
5      7.90       156.46     A        A4  ...     NaN         NaN         NaN   
6     15.96       170.08     C        C5  ...     NaN         NaN         NaN   
7     18.64       109.43     E        E1  ...     NaN         NaN         NaN   
8     21.28       152.39     F        F2  ...     NaN         NaN         NaN   
9     12.69       121.45     B        B5  ...     NaN         NaN         NaN   

   max_bal_bc all_util total_rev_hi_lim inq_fi total_cu_tl inq_last_12m  \
0         NaN      NaN              NaN    NaN         NaN          NaN   
1         NaN      NaN              NaN    NaN         NaN          NaN   
2         NaN      NaN              NaN    NaN         NaN          NaN   
3         NaN      NaN              NaN    NaN         NaN          NaN   
4         NaN      NaN              NaN    NaN         NaN          NaN   
5         NaN      NaN              NaN    NaN         NaN          NaN   
6         NaN      NaN              NaN    NaN         NaN          NaN   
7         NaN      NaN              NaN    NaN         NaN          NaN   
8         NaN      NaN              NaN    NaN         NaN          NaN   
9         NaN      NaN              NaN    NaN         NaN          NaN   

  default_ind  
0           0  
1           1  
2           0  
3           0  
4           0  
5           0  
6           0  
7           0  
8           1  
9           1  

[10 rows x 73 columns]

To advance to the data analysis we need to gather the test and the train samples. To do so we will import the function train_test_split from sklearn and use it on the credit_data data.

In [113]:
from sklearn import neighbors, datasets, preprocessing
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

X, y = credit_data[credit_data.columns.drop('default_ind')], credit_data['default_ind']
    
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.01, random_state=33)

X_test.head(10)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
88331    7081256    8743009      11500        11500          11500.0   
577572  60336606   64313389      25000        25000          25000.0   
167966   3066805    3739422       5375         5375           5375.0   
601264  59130093   63017845       6000         6000           6000.0   
404061  13317737   15359946      26000        26000          26000.0   
560190  61371268   65490059      12000        12000          12000.0   
705487  50173398   53552116      12000        12000          12000.0   
640554  55950785   59612595       3000         3000           3000.0   
707318  50615917   53995663      16000        16000          15975.0   
629611  57214309   60917030      12500        12500          12500.0   

              term  int_rate  installment grade sub_grade  ... total_bal_il  \
88331    36 months     15.88       403.63     C        C4  ...          NaN   
577572   36 months      6.89       770.68     A        A3  ...          NaN   
167966   36 months     10.16       173.85     B        B1  ...          NaN   
601264   36 months      7.89       187.72     A        A5  ...          NaN   
404061   60 months     14.16       607.14     C        C2  ...          NaN   
560190   60 months     15.61       289.34     D        D1  ...          NaN   
705487   36 months      6.24       366.37     A        A2  ...          NaN   
640554   36 months     10.99        98.21     B        B4  ...          NaN   
707318   60 months     19.99       423.82     E        E4  ...          NaN   
629611   36 months      6.24       381.64     A        A2  ...          NaN   

       il_util open_rv_12m  open_rv_24m max_bal_bc all_util total_rev_hi_lim  \
88331      NaN         NaN          NaN        NaN      NaN          17400.0   
577572     NaN         NaN          NaN        NaN      NaN          46600.0   
167966     NaN         NaN          NaN        NaN      NaN          22200.0   
601264     NaN         NaN          NaN        NaN      NaN          25000.0   
404061     NaN         NaN          NaN        NaN      NaN          79200.0   
560190     NaN         NaN          NaN        NaN      NaN          13100.0   
705487     NaN         NaN          NaN        NaN      NaN          80800.0   
640554     NaN         NaN          NaN        NaN      NaN          75200.0   
707318     NaN         NaN          NaN        NaN      NaN          10800.0   
629611     NaN         NaN          NaN        NaN      NaN          68600.0   

       inq_fi total_cu_tl inq_last_12m  
88331     NaN         NaN          NaN  
577572    NaN         NaN          NaN  
167966    NaN         NaN          NaN  
601264    NaN         NaN          NaN  
404061    NaN         NaN          NaN  
560190    NaN         NaN          NaN  
705487    NaN         NaN          NaN  
640554    NaN         NaN          NaN  
707318    NaN         NaN          NaN  
629611    NaN         NaN          NaN  

[10 rows x 72 columns]

This way, the test sample will have results with a size corresponding to 1% of the total data.

In [ ]:
credit_data.tail()

In [4]:
credit_data.describe()

id     member_id      loan_amnt    funded_amnt  \
count  8.559690e+05  8.559690e+05  855969.000000  855969.000000   
mean   3.224073e+07  3.476269e+07   14745.571335   14732.378305   
std    2.271969e+07  2.399418e+07    8425.340005    8419.471653   
min    5.473400e+04  7.069900e+04     500.000000     500.000000   
25%    9.067986e+06  1.079273e+07    8000.000000    8000.000000   
50%    3.431355e+07  3.697532e+07   13000.000000   13000.000000   
75%    5.446311e+07  5.803559e+07   20000.000000   20000.000000   
max    6.861687e+07  7.351969e+07   35000.000000   35000.000000   

       funded_amnt_inv       int_rate    installment    annual_inc  \
count    855969.000000  855969.000000  855969.000000  8.559690e+05   
mean      14700.061226      13.192320     436.238072  7.507119e+04   
std        8425.805478       4.368365     243.726876  6.426447e+04   
min           0.000000       5.320000      15.690000  0.000000e+00   
25%        8000.000000       9.990000     260.550000  4.500000e+04   
50%       13000.000000      12.990000     382.550000  6.500000e+04   
75%       20000.000000      15.990000     571.560000  9.000000e+04   
max       35000.000000      28.990000    1445.460000  9.500000e+06   

                 dti    delinq_2yrs  ...       il_util   open_rv_12m  \
count  855969.000000  855969.000000  ...  11609.000000  13288.000000   
mean       18.122165       0.311621  ...     71.486993      1.354305   
std        17.423629       0.857189  ...     23.015293      1.483710   
min         0.000000       0.000000  ...      0.000000      0.000000   
25%        11.880000       0.000000  ...     58.500000      0.000000   
50%        17.610000       0.000000  ...     75.000000      1.000000   
75%        23.900000       0.000000  ...     87.500000      2.000000   
max      9999.000000      39.000000  ...    223.300000     22.000000   

        open_rv_24m    max_bal_bc      all_util  total_rev_hi_lim  \
count  13288.000000  13288.000000  13288.000000      7.886560e+05   
mean       2.945515   5840.443332     61.024526      3.216357e+04   
std        2.595313   5108.500262     20.018117      3.769964e+04   
min        0.000000      0.000000      0.000000      0.000000e+00   
25%        1.000000   2405.000000     47.900000      1.400000e+04   
50%        2.000000   4485.500000     62.100000      2.380000e+04   
75%        4.000000   7701.250000     75.300000      3.990000e+04   
max       43.000000  83047.000000    151.400000      9.999999e+06   

             inq_fi   total_cu_tl  inq_last_12m    default_ind  
count  13288.000000  13288.000000  13288.000000  855969.000000  
mean       0.947772      1.524232      1.841963       0.054286  
std        1.441667      2.697601      2.975049       0.226581  
min        0.000000      0.000000     -4.000000       0.000000  
25%        0.000000      0.000000      0.000000       0.000000  
50%        0.000000      0.000000      2.000000       0.000000  
75%        1.000000      2.000000      3.000000       0.000000  
max       15.000000     33.000000     32.000000       1.000000  

[8 rows x 52 columns]